File: frontend/src/types/evaluations.ts
export type RagEvaluationJobStatus =
  'queued' | 'running' | 'cancelling' | 'cancelled' | 'succeeded' | 'failed';

export type RagEvaluationCaseStatus = 'queued' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface RagExpectedBehavior {
  complexity?: 'simple' | 'complex' | null;
  route?: string | null;
  outcome?: 'ANSWERABLE' | 'NO_KNOWLEDGE' | 'INSUFFICIENT_EVIDENCE' | null;
  hitl?: 'none' | 'clarify' | 'scope_select';
  acceptable_abstention?: boolean;
  hitl_resolution_success?: boolean | null;
  hitl_final_outcome?: string | null;
}

export interface RagGoldDocument {
  document_id?: string | null;
  canonical_name?: string | null;
}

export interface RagGoldChunk {
  chunk_id?: string | null;
  content_sha256?: string | null;
}

export interface RagEvaluationDatasetCase {
  id: string;
  tags?: string[];
  critical?: boolean;
  question: string;
  expected: RagExpectedBehavior;
  gold_documents?: RagGoldDocument[];
  gold_chunks?: RagGoldChunk[];
  reference_answer?: string | null;
  required_claims?: string[];
  conflicts?: string[];
  hitl_answers?: string[];
}

export interface RagEvaluationDataset {
  schema_version: 1;
  name: string;
  cases: RagEvaluationDatasetCase[];
}

export interface RagEvaluationDatasetRecord {
  id: string;
  name: string;
  fingerprint: string;
  case_count: number;
  dataset: RagEvaluationDataset;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface RagMetricGate {
  metric: string;
  direction: 'higher_is_better' | 'lower_is_better';
  minimum?: number | null;
  maximum?: number | null;
  max_regression?: number;
  required?: boolean;
}

export interface RagEvaluationGatePolicy {
  schema_version?: 1;
  k_values?: number[];
  critical_no_regression?: boolean;
  required_provenance?: 'contract_smoke' | 'live_rag' | null;
  metric_gates?: RagMetricGate[];
}

export interface EvaluationModelSummary {
  profile_id: string;
  profile_version: number;
  display_name: string;
  provider: string;
  model_name: string;
  timeout_seconds: number;
  supports_stream: boolean;
  supports_structured_output: boolean;
}

export interface RagMetricResult {
  value: number | null;
  eligible_cases: number;
}

export interface RagGateResult {
  name: string;
  status: 'passed' | 'failed' | 'skipped';
  metric?: string | null;
  actual?: number | null;
  baseline?: number | null;
  threshold?: number | null;
  baseline_threshold?: number | null;
  detail: string;
}

export interface RagReportCase {
  case_id: string;
  critical: boolean;
  metrics: Record<string, number | null>;
  checks: Record<string, boolean | null>;
  provider_failed: boolean;
  provider_error_code?: string | null;
  provider_error_stage?: 'retrieval' | 'generation' | 'judge' | null;
  gold_chunk_count: number;
  matched_gold_chunk_count: number;
  passed: boolean;
}

export interface RagEvaluationReport {
  schema_version: 1;
  dataset_name: string;
  dataset_fingerprint: string;
  case_count: number;
  observation_count: number;
  metrics: Record<string, RagMetricResult>;
  slices: Record<string, { case_count: number; metrics: Record<string, RagMetricResult> }>;
  unavailable_metrics: Record<string, string>;
  cases: RagReportCase[];
  gates: RagGateResult[];
  passed: boolean;
  metadata: Record<string, unknown>;
}

export interface RagEvaluationJob {
  id: string;
  dataset_id: string;
  dataset_name: string;
  dataset_fingerprint: string;
  baseline_job_id: string | null;
  status: RagEvaluationJobStatus;
  completed_cases: number;
  total_cases: number;
  progress: number;
  gate_policy: RagEvaluationGatePolicy;
  model_catalog_hash: string;
  models: Record<string, EvaluationModelSummary>;
  owner_worker_id: string | null;
  lease_expires_at: string | null;
  fencing_token: number;
  attempts: number;
  max_attempts: number;
  error_code: string | null;
  error: Record<string, unknown> | null;
  report: RagEvaluationReport | null;
  created_by: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RagJudgeMetrics {
  answer_correctness: number;
  groundedness: number;
  answer_relevance: number;
  completeness: number;
  context_relevance: number;
  unsupported_claim_rate: number;
  conflict_disclosure_rate: number;
  reason?: string;
}

export interface RagEvaluationObservation {
  case_id: string;
  complexity?: string | null;
  route?: string | null;
  outcome?: string | null;
  hitl?: string;
  rewrite_performed?: boolean;
  provider_error_code?: string | null;
  provider_error_stage?: 'retrieval' | 'generation' | 'judge' | null;
  duration_ms: number;
  judge?: RagJudgeMetrics | null;
}

export interface RagEvaluationCaseResult {
  id: string;
  job_id: string;
  case_id: string;
  position: number;
  status: RagEvaluationCaseStatus;
  question: string;
  generated_answer: string | null;
  judge_reason: string | null;
  observation: RagEvaluationObservation | null;
  judge: RagJudgeMetrics | null;
  metrics: Record<string, number | null>;
  checks: Record<string, boolean | null>;
  retrieved_identities: Array<Record<string, unknown>>;
  provider_error_code: string | null;
  provider_error_stage?: 'retrieval' | 'generation' | 'judge' | null;
  duration_ms: number | null;
  error_code: string | null;
  error: Record<string, unknown> | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RagEvaluationJobCreatePayload {
  dataset_id: string;
  baseline_job_id?: string | null;
  gate_policy?: RagEvaluationGatePolicy | null;
}

File: backend/schemas/evaluations.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from backend.evaluation.contracts import (
    RagEvaluationCaseStatus,
    RagEvaluationJobStatus,
)
from backend.evaluation.rag import (
    RagEvalDataset,
    RagEvalGatePolicy,
    RagEvalObservation,
    RagEvalReport,
)


class EvaluationSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RagEvaluationDatasetCreateRequest(EvaluationSchema):
    dataset: RagEvalDataset


class RagEvaluationDatasetResponse(EvaluationSchema):
    id: str
    name: str
    fingerprint: str
    case_count: int
    dataset: RagEvalDataset
    created_by: str | None = None
    created_at: datetime
    updated_at: datetime


class RagEvaluationDatasetsResponse(EvaluationSchema):
    datasets: tuple[RagEvaluationDatasetResponse, ...]


class RagEvaluationJobCreateRequest(EvaluationSchema):
    dataset_id: str = Field(min_length=1, max_length=64)
    baseline_job_id: str | None = Field(default=None, max_length=64)
    gate_policy: RagEvalGatePolicy | None = None


class EvaluationModelSummary(EvaluationSchema):
    profile_id: str
    profile_version: int
    display_name: str
    provider: str
    model_name: str
    timeout_seconds: float
    supports_stream: bool
    supports_structured_output: bool


class RagEvaluationJobResponse(EvaluationSchema):
    id: str
    dataset_id: str
    dataset_name: str
    dataset_fingerprint: str
    baseline_job_id: str | None = None
    status: RagEvaluationJobStatus
    completed_cases: int
    total_cases: int
    progress: float = Field(ge=0, le=1)
    gate_policy: RagEvalGatePolicy
    model_catalog_hash: str
    models: dict[str, EvaluationModelSummary]
    owner_worker_id: str | None = None
    lease_expires_at: datetime | None = None
    fencing_token: int
    attempts: int
    max_attempts: int
    error_code: str | None = None
    error: dict | None = None
    report: RagEvalReport | None = None
    created_by: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class RagEvaluationJobsResponse(EvaluationSchema):
    jobs: tuple[RagEvaluationJobResponse, ...]


class RagEvaluationCaseResponse(EvaluationSchema):
    id: str
    job_id: str
    case_id: str
    position: int
    status: RagEvaluationCaseStatus
    question: str
    generated_answer: str | None = None
    judge_reason: str | None = None
    observation: RagEvalObservation | None = None
    judge: dict | None = None
    metrics: dict[str, float | None]
    checks: dict[str, bool | None]
    retrieved_identities: tuple[dict, ...]
    provider_error_code: str | None = None
    provider_error_stage: str | None = None
    duration_ms: int | None = None
    error_code: str | None = None
    error: dict | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class RagEvaluationCasesResponse(EvaluationSchema):
    cases: tuple[RagEvaluationCaseResponse, ...]


__all__ = [
    "EvaluationModelSummary",
    "RagEvaluationCaseResponse",
    "RagEvaluationCasesResponse",
    "RagEvaluationDatasetCreateRequest",
    "RagEvaluationDatasetResponse",
    "RagEvaluationDatasetsResponse",
    "RagEvaluationJobCreateRequest",
    "RagEvaluationJobResponse",
    "RagEvaluationJobsResponse",
]
